### PR TITLE
follow-up work on not visible indexes

### PR DESCRIPTION
**Pull request by taroface**
Fixes [DOC-4743](https://cockroachlabs.atlassian.net/browse/DOC-4743).

This feature was largely documented with https://github.com/cockroachdb/docs/pull/15090. This PR adds some refinements to how & where it is described, and specifies that `UNIQUE` and `PRIMARY KEY` constraints are still respected on not visible indexes.

### CODE REVIEW

**Review by rytaft — APPROVED** [review]
:lgtm: just some nits

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15306#-:-NDj7lFi4YtquJdZhDsl:b-yl2rbz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dikshant, @mgartner, @michae2, and @taroface)

___
*[`v22.2/alter-index.md` line 28 at r1](https://reviewable.io/reviews/cockroachdb/docs/15306#-NDj57Ak3-gvs2meQjOE:-NDj57Ak3-gvs2meQjOF:b-h2z83i) ([raw file](https://github.com/cockroachdb/docs/blob/d968544a03ad4fb35eca157cb7da181d5da5b1e6/v22.2/alter-index.md#L28)):*
> ```Markdown
> Use the `VISIBLE` and `NOT VISIBLE` subcommands to set the visibility of an index. This determines whether the index is visible to the [cost-based optimizer](cost-based-optimizer.html#control-whether-the-optimizer-uses-an-index). 
> 
> By default, indexes are visible. If `NOT VISIBLE`, the index will not be used in queries unless it is specifically selected with [index hint](indexes.html#selection) or the property is overridden with the [`optimizer_use_not_visible_indexes` session variable](set-vars.html#optimizer-use-not-visible-indexes).
> ```

nit: with [index hint] -> with an [index hint]
___
*[`v22.2/create-index.md` line 56 at r1](https://reviewable.io/reviews/cockroachdb/docs/15306#-NDj5HXJDUqbwJY_hKnf:-NDj5HXJDUqbwJY_hKng:b-h2z83i) ([raw file](https://github.com/cockroachdb/docs/blob/d968544a03ad4fb35eca157cb7da181d5da5b1e6/v22.2/create-index.md#L56)):*
> ```Markdown
> `opt_partition_by` | An [Enterprise-only](enterprise-licensing.html) option that lets you [define index partitions at the row level](partitioning.html). As of CockroachDB v21.1 and later, most users should use [`REGIONAL BY ROW` tables](multiregion-overview.html#regional-by-row-tables). Indexes against regional by row tables are automatically partitioned, so explicit index partitioning is not required.
> `opt_where_clause` |  An optional `WHERE` clause that defines the predicate boolean expression of a [partial index](partial-indexes.html).
> `opt_index_visible` | An optional `VISIBLE` or `NOT VISIBLE` clause that indicates whether an index is visible to the [cost-based optimizer](cost-based-optimizer.html#control-whether-the-optimizer-uses-an-index). If `NOT VISIBLE`, the index will not be used in queries unless it is specifically selected with [index hint](indexes.html#selection) or the property is overridden with the [`optimizer_use_not_visible_indexes` session variable](set-vars.html#optimizer-use-not-visible-indexes). For an example, see [Set an index to be not visible](alter-index.html#set-an-index-to-be-not-visible).<br><br>`UNIQUE` and `PRIMARY KEY` [constraints](constraints.html) are still enforced on indexes that are not visible.
> ```

nit: with [index hint] -> with an [index hint]
___
*[`v22.2/create-table.md` line 112 at r1](https://reviewable.io/reviews/cockroachdb/docs/15306#-NDj5M-nB4ouxn4za-4V:-NDj5M-nB4ouxn4za-4W:b-h2z83i) ([raw file](https://github.com/cockroachdb/docs/blob/d968544a03ad4fb35eca157cb7da181d5da5b1e6/v22.2/create-table.md#L112)):*
> ```Markdown
> `opt_locality` |  Specify a [locality](multiregion-overview.html#table-locality) for the table. In order to set a locality, the table must belong to a [multi-region database](multiregion-overview.html).<br><br>Note that multi-region features require an [Enterprise license](enterprise-licensing.html).
> `opt_where_clause` | An optional `WHERE` clause that defines the predicate boolean expression of a [partial index](partial-indexes.html).
> `opt_index_visible` | An optional `VISIBLE` or `NOT VISIBLE` clause that indicates whether an index is visible to the [cost-based optimizer](cost-based-optimizer.html#control-whether-the-optimizer-uses-an-index). If `NOT VISIBLE`, the index will not be used in queries unless it is specifically selected with [index hint](indexes.html#selection) or the property is overridden with the [`optimizer_use_not_visible_indexes` session variable](set-vars.html#optimizer-use-not-visible-indexes). For an example, see [Set an index to be not visible](alter-index.html#set-an-index-to-be-not-visible).<br><br>`UNIQUE` and `PRIMARY KEY` [constraints](constraints.html) are still enforced on indexes that are not visible.
> ```

nit: with [index hint] -> with an [index hint]
___
*[`v22.2/indexes.md` line 61 at r1](https://reviewable.io/reviews/cockroachdb/docs/15306#-NDj7aKV031UPYUobsT8:-NDj7aKV031UPYUobsT9:b-h2z83i) ([raw file](https://github.com/cockroachdb/docs/blob/d968544a03ad4fb35eca157cb7da181d5da5b1e6/v22.2/indexes.md#L61)):*
> ```Markdown
> To maximize your indexes' performance, Cockroach Labs recommends following the [secondary index best practices](schema-design-indexes.html#best-practices).
> 
> To observe the impact of an index without affecting a production workload, you can [create an index](create-index.html) using the `NOT VISIBLE` clause. The index will not be used in queries unless it is specifically selected with [index hint](indexes.html#selection) or the property is overridden with the [`optimizer_use_not_visible_indexes` session variable](set-vars.html#optimizer-use-not-visible-indexes). For an example, see [Set an index to be not visible](alter-index.html#set-an-index-to-be-not-visible).
> ```

nit: with [index hint] -> with an [index hint]

**Review by michae2 — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15306)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dikshant, @kathancox, @mgartner, @rytaft, and @taroface)

___
*[`v22.2/alter-index.md` line 32 at r2](https://reviewable.io/reviews/cockroachdb/docs/15306#-NDjOlDTC_7NTjV2u7yf:-NDjOlDTC_7NTjV2u7yg:b315yt) ([raw file](https://github.com/cockroachdb/docs/blob/f221f22a42248caa76f95502f47f90ab25028bc5/v22.2/alter-index.md#L32)):*
> ```Markdown
> {{site.data.alerts.callout_info}}
> `UNIQUE` and `PRIMARY KEY` [constraints](constraints.html) are still enforced on indexes that are not visible.
> {{site.data.alerts.end}}
> ```

More caveats:
* Primary indexes cannot be not visible / must be visible.
* Indexes that are not visible are still used for foreign key cascades.
* Indexes that are not visible are still used to enforce foreign key constraints.
* The `NOT VISIBLE` syntax works with unique indexes, but not with unique constraints. To make the indexes for unique constraints not visible, `ALTER INDEX` must be used after creating the unique constraint.

**Comment by taroface** [comment]
@michae2 Thank you! I rephrased a bit; please let me know if this is still accurate. Also, I had erroneously mentioned primary key instead of foreign key constraints, so I removed the reference to the former (not sure if primary key constraint also applies and/or is worth keeping?).

**Review by michae2 — APPROVED** [review]
:lgtm: This looks great! I think removing the reference to primary key constraints was good. Thank you!

> Reviewed 2 of 7 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15306#-:-NE7xlOuAFM2JYGtmGQg:b-qz235x)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dikshant, @mgartner, @rytaft, and @taroface)

___
*[`v22.2/indexes.md` line 61 at r3](https://reviewable.io/reviews/cockroachdb/docs/15306#-NE7xPCTEiUSAo6o1FwW:-NE7xPCUA0p2M_6jDfmG:b9lq9i5) ([raw file](https://github.com/cockroachdb/docs/blob/769244138f037a45b809e87ed6f87036657c026c/v22.2/indexes.md#L61)):*
> ```Markdown
> To maximize your indexes' performance, Cockroach Labs recommends following the [secondary index best practices](schema-design-indexes.html#best-practices).
> 
> To observe the impact of an index without affecting a production workload, you can [create an index](create-index.html) using the `NOT VISIBLE` clause. The index will not be used in queries unless it is specifically selected with an [index hint](indexes.html#selection) or the property is overridden with the [`optimizer_use_not_visible_indexes` session variable](set-vars.html#optimizer-use-not-visible-indexes). For an example, see [Set an index to be not visible](alter-index.html#set-an-index-to-be-not-visible).
> ```

nit: Might be nice to also mention `UNIQUE` and `FOREIGN KEY` constraints here, or at least link to the `Index visibility considerations` section.

**Review by kathancox — APPROVED** [review]
LGTM, only one very optional comment!

File: v22.2/cost-based-optimizer.md [hunk 1]
```diff
@@ -192,7 +192,13 @@ If you disable full scans, and you provide an [index hint](indexes.html#selectio
 
 ## Control whether the optimizer uses an index
 
-You can specify [whether an index is visible](alter-index.html#set-an-index-to-be-not-visible) to the cost-based optimizer. If not visible, the index will not be used in queries unless specifically selected with an [index hint](indexes.html#selection). This allows you to create an index and check for query plan changes without affecting production queries. For an example, see [Set an index to be not visible](alter-index.html#set-an-index-to-be-not-visible).
+You can specify [whether an index is visible](alter-index.html#index-visibility) to the cost-based optimizer. By default, indexes are visible. If not visible, the index will not be used in queries unless it is specifically selected with an [index hint](indexes.html#selection).
```
[the remark this answers is not in the document]
Comment (taroface) [reply v22.2/cost-based-optimizer.md:195]: Hmm I'm not sure if the optimizer can be said to handle queries beyond computing the plan. That is, using active voice here may ascribe more 'agency' to the optimizer than it has, although I'm not sure. I'll learn more about this and keep it in mind for future work!

[the remark this answers is not in the document]
Comment (kathancox) [reply v22.2/cost-based-optimizer.md:195]: Sure, SGTM, I don't have knowledge here!

[the remark this answers is not in the document]
Comment (mgartner) [reply v22.2/cost-based-optimizer.md:195]: I think @kathancox's suggest would be understood by most readers, though I agree with @taroface that it's technically not the optimizer that is _using_ an index or not. I tend to use the verbs "considers" or "explores" when writing about the optimizer behavior, for example:

"If not visible, the cost-based optimizer will not consider the index when planning queries..."

"If not visible, the cost-based optimizer will not explore query plans that use the index unless..."

Though I don't feel that those are more clear than what's been merged, so no need to change anything.

[the remark this answers is not in the document]
Comment (taroface) [reply v22.2/cost-based-optimizer.md:195]: This is very helpful! Thank you, and I'll use these in the future.

**Pull request merged**